python-oauth-web-app/atproto_oauth.py

```diff
@@ -0,0 +1,256 @@
+        {
+            "iss": client_id,
+            "sub": client_id,
+            "aud": authsrv_url,
+            "jti": generate_token(),
+            "iat": int(time.time()),
+        },
+        secret_jwk,
+    ).decode("utf-8")
+
+    par_body = {
+        "response_type": "code",
+        "code_challenge": code_challenge,
+        "code_challenge_method": code_challenge_method,
+        "client_id": client_id,
+        "state": state,
+        "nonce": nonce,
+        "redirect_uri": redirect_uri,
+        "scope": scope,
+        "login_hint": username,
+        "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+        "client_assertion": client_assertion,
+    }
+    # print(par_body)
+
+    # TODO: could be doing DPoP here
+    # IMPORTANT: Pushed Authorization Request URL is untrusted input, SSRF mitigations are needed
+    assert is_safe_url(par_url)
+    with hardened_http.get_session() as sess:
+        resp = sess.post(
+            par_url,
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+            data=par_body,
+        )
+    return pkce_verifier, state, nonce, resp
+
+
+def complete_auth_request(auth_request, code, app_url, secret_jwk):
+
+    state = auth_request["state"]
+    authsrv_url = auth_request["iss"]
+
+    # Re-fetch server metadata
+    authsrv_meta = fetch_authsrv_meta(authsrv_url)
+
+    # Construct auth token request fields
+    client_id = f"{app_url}oauth/client-metadata.json"
+    redirect_uri = f"{app_url}oauth/callback"
+
+    # This is where the "confidential client" secret key is used
+    client_assertion = jwt.encode(
+        {"alg": "ES256", "kid": secret_jwk["kid"]},
+        {
+            "iss": client_id,
+            "sub": client_id,
+            "aud": authsrv_url,
+            "jti": generate_token(),
+            "iat": int(time.time()),
+        },
+        secret_jwk,
+    ).decode("utf-8")
+
+    params = {
+        "client_id": client_id,
+        "redirect_uri": redirect_uri,
+        "grant_type": "authorization_code",
+        "code": code,
+        "code_verifier": auth_request["pkce_verifier"],
+        "client_assertion_type": "urn:ietf:params:oauth:client-assertion-type:jwt-bearer",
+        "client_assertion": client_assertion,
+    }
+
+    # Create DPoP header JWT, using the existing DPoP signing key for this account/session
+    token_url = authsrv_meta["token_endpoint"]
+    dpop_key = JsonWebKey.import_key(json.loads(auth_request["dpop_private_jwk"]))
+    dpop_pub_jwk = json.loads(dpop_key.as_json(is_private=False))
+    dpop_proof = jwt.encode(
+        {"typ": "dpop+jwt", "alg": "ES256", "jwk": dpop_pub_jwk},
+        {
+            "jti": generate_token(),
+            "htm": "POST",
+            "htu": token_url,
+            "iat": int(time.time()),
+            "exp": int(time.time()) + 30,
+        },
+        dpop_key,
+    ).decode("utf-8")
+
+    # TODO: should we be reusing earlier DPoP nonce?
```

Comment:
I'm interested in the answer to this!

Reply:
It is a good practice to keep a local map between origin (`https://my-resource-server`) and previously issued nonce. It is also a good practice to update that local map using the `DPoP-nonce` header of every HTTP response.

Reply:
(sort answer is "yes")

Reply:
updated to be more consistent about nonce naming and storage